src/community/jdbcconfig/src/main/java/org/geoserver/jdbcconfig/internal/ConfigDatabase.java

```diff
@@ -1678,4 +1692,110 @@ public void handleSettingsAdded(SettingsInfo settings) {
             updateCache(settings);
         }
     }
+
+    public class CatalogReferenceUpdater implements CatalogVisitor {
+
+        private CatalogReferenceUpdater() {}
+
+        @Override
+        public void visit(Catalog catalog) {}
+
+        @Override
+        public void visit(WorkspaceInfo workspace) {}
+
+        @Override
+        public void visit(NamespaceInfo workspace) {}
+
+        @Override
+        public void visit(DataStoreInfo dataStore) {
+            if (dataStore.getWorkspace() != null) {
+                dataStore.setWorkspace(
+                        getById(dataStore.getWorkspace().getId(), WorkspaceInfo.class));
+            }
+        }
+
+        @Override
+        public void visit(CoverageStoreInfo coverageStore) {
+            if (coverageStore.getWorkspace() != null) {
+                coverageStore.setWorkspace(
+                        getById(coverageStore.getWorkspace().getId(), WorkspaceInfo.class));
+            }
+        }
+
+        @Override
+        public void visit(WMSStoreInfo wmsStore) {
+            if (wmsStore.getWorkspace() != null) {
+                wmsStore.setWorkspace(
+                        getById(wmsStore.getWorkspace().getId(), WorkspaceInfo.class));
+            }
+        }
+
+        @Override
+        public void visit(FeatureTypeInfo featureType) {
+            if (featureType.getNamespace() != null) {
+                featureType.setNamespace(
+                        getById(featureType.getNamespace().getId(), NamespaceInfo.class));
+            }
+            featureType.setStore(getById(featureType.getStore().getId(), StoreInfo.class));
+        }
+
+        @Override
+        public void visit(CoverageInfo coverage) {
+            if (coverage.getNamespace() != null) {
+                coverage.setNamespace(
+                        getById(coverage.getNamespace().getId(), NamespaceInfo.class));
+            }
+            coverage.setStore(getById(coverage.getStore().getId(), StoreInfo.class));
+        }
+
+        @Override
+        public void visit(LayerInfo layer) {
+            if (layer.getDefaultStyle() != null) {
+                layer.setDefaultStyle(getById(layer.getDefaultStyle().getId(), StyleInfo.class));
+            }
+            Set<StyleInfo> newStyles = new HashSet<>();
+            for (StyleInfo style : layer.getStyles()) {
+                if (style != null) {
+                    newStyles.add(getById(style.getId(), StyleInfo.class));
+                }
+            }
+            layer.getStyles().clear();
+            layer.getStyles().addAll(newStyles);
+        }
+
+        @Override
+        public void visit(StyleInfo style) {
+            if (style.getWorkspace() != null) {
+                style.setWorkspace(getById(style.getWorkspace().getId(), WorkspaceInfo.class));
+            }
+        }
+
+        @Override
+        public void visit(LayerGroupInfo layerGroup) {
```

Comment:
Do the RootLayer, RootLayerStyle and Styles also need to be updated?

Reply:
indeed

Reply:
commit added for this

Reply:
LayerGroupInfo styles are a List (not a Set like LayerInfo) and the order of the styles does matter.

Reply:
Yes, indeed. Fixed this.